01.Importing Libraries 

In [107]:
#import libraries 
import pandas as pd
import numpy as np
import os

02.Importing Data

In [109]:
#create path
path = r"C:\Users\nabar\OneDrive\Documents\03-2025 Instacart Basket Analysis"

In [110]:
#products data
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original data', 'products.csv'), index_col = False)

In [111]:
# orders data
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data','orders_wrangled.csv'), index_col = False)

In [112]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number     order_dow  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01  2.776219e+00   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01  2.046829e+00   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00  1.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01  3.000000e+00   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01  5.000000e+00   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02  6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

03.Mixed Type Data

In [114]:
#create data frame
df_test = pd.DataFrame()

In [115]:
#create a mixed type column
df_test['mix'] = ['a', 'b', 1, True]

In [116]:
df_test. head()

mix
0     a
1     b
2     1
3  True

In [117]:
#check for mixed data type
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


C:\Users\nabar\AppData\Local\Temp\ipykernel_10536\26977339.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)


In [118]:
#check for mixed data type
for col in df_test.columns.tolist():
  weird = (df_test[[col]].map(type) != df_test[[col]].iloc[0].map(type)).any(axis=1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [119]:
#coverting mixed data type to string
df_test['mix'] = df_test['mix'].astype('str')

04.Missing Values

In [121]:
#Finding missing values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [122]:
##Creating a datafarme to subset the data in the df_prods where there  missing values are found  in the product_name column
df_nan=df_prods[df_prods['product_name'].isnull()==True]

In [123]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [124]:
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

In [125]:
df_prods.shape

(49693, 5)

In [126]:
#Creating a new dataframe as a subset of df_prods dataframe where the missing values in the product_name variable are excluded
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [127]:
df_prods_clean.shape

(49677, 5)

In [128]:
df_prods.dropna(subset = ['product_name'], inplace = True)

In [129]:
df_prods.shape

(49677, 5)

05.Duplicate Values

In [131]:
#Finding duplicates
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [132]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [133]:
df_prods_clean.shape

(49677, 5)

In [134]:
#Creating a new datafrme where there is no duplicate values
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [135]:
df_prods_clean_no_dups.shape

(49672, 5)

In [136]:
#Export the cleaned datafrme for df_prods
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

06.Insight
Order Data Summary:

Customers place orders approximately every 11 days on average, mostly in the early afternoon, with peak activity earlier in the week. There is considerable variation in ordering habits.

Missing Data in Product Names:

A small number of rows (16) have missing product names, which might affect product-level analyses. Other columns like aisle_id and department_id remain intact.

Data Cleaning Impact:

After removing rows with missing values, the cleaned product dataset is reduced to 49,677 rows. This ensures accurate product insights.

Duplicate Product Entries:

5 duplicate rows were identified and removed from the cleaned product dataset. This refinement leaves 49,672 unique entries.

In [138]:
# Running describe function in the df_ords dataframe
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number     order_dow  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01  2.776219e+00   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01  2.046829e+00   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00  1.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01  3.000000e+00   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01  5.000000e+00   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02  6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

Nothing extraordinery found here.

In [140]:
#check for mixed data type
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)  

C:\Users\nabar\AppData\Local\Temp\ipykernel_10536\841247415.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\nabar\AppData\Local\Temp\ipykernel_10536\841247415.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\nabar\AppData\Local\Temp\ipykernel_10536\841247415.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\nabar\AppData\Local\Temp\ipykernel_10536\841247415.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\nabar\AppData\Local\Tem

No mixed data type found in any columns

In [142]:
#Checking the data type of all columns
df_ords.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3421083 entries, 0 to 3421082
Data columns (total 8 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   Unnamed: 0              int64  
 1   order_id                int64  
 2   user_id                 int64  
 3   eval_set                object 
 4   order_number            int64  
 5   order_dow               int64  
 6   order_hour_of_day       int64  
 7   days_since_prior_order  float64
dtypes: float64(1), int64(6), object(1)
memory usage: 208.8+ MB


There is no mixed-type data found in any of the columns.

In [144]:
#Checking missing values 
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
eval_set                       0
order_number                   0
order_dow                      0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

In [145]:
df_ords.shape

(3421083, 8)

Findings:
The column days_since_prior_order has 206,209 missing values, which likely correspond to first-time orders made by users. When a user places their first order, there is no prior order to calculate the gap, so the value is recorded as missing. This makes sense and provides meaningful information about user behavior.
To preserve the integrity of the data and maintain insights about customer order patterns, the missing values will be kept. These entries can later be labeled as "First Order" for easier analysis and interpretation. By keeping them, the dataset remains logical and useful for further exploration.




In [147]:
#Finding duplicates 
df_ords[df_ords.duplicated()]

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, eval_set, order_number, order_dow, order_hour_of_day, days_since_prior_order]
Index: []

There are no duplicate values found in the df_ords.

In [149]:
#Dropping the column "Unnamed: 0" from the df_ords dataframe and assinged it to a new dataframe "df_ords_clean"
df_ords_clean = df_ords.drop('Unnamed: 0', axis=1)

In [150]:
df_ords_clean.head()

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329        1    prior             1          2                  8   
1   2398795        1    prior             2          3                  7   
2    473747        1    prior             3          3                 12   
3   2254736        1    prior             4          4                  7   
4    431534        1    prior             5          4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

Found unnamed column in df_ords. Dropped column 'unnamed'.

In [201]:
df_ords_clean.rename(columns = {'order_dow' : 'orders_day_of_the_week'}, inplace=True)

In [203]:
df_ords_clean.head()


order_id  user_id eval_set  order_number  orders_day_of_the_week  \
0   2539329        1    prior             1                       2   
1   2398795        1    prior             2                       3   
2    473747        1    prior             3                       3   
3   2254736        1    prior             4                       4   
4    431534        1    prior             5                       4   

   order_hour_of_day  days_since_prior_order  
0                  8                     NaN  
1                  7                    15.0  
2                 12                    21.0  
3                  7                    29.0  
4                 15                    28.0

In [205]:
df_ords_clean.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_checked.csv'), index=False)